// serverless-websockets-plugin/dist/index.js
"use strict";

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var _require = require('@serverless/utils'),
    is = _require.is,
    map = _require.map,
    all = _require.all,
    filter = _require.filter,
    keys = _require.keys,
    isEmpty = _require.isEmpty,
    flatten = _require.flatten;

var chalk = require('chalk');

var ServerlessWebsocketsPlugin =
/*#__PURE__*/
function () {
  function ServerlessWebsocketsPlugin(serverless, options) {
    _classCallCheck(this, ServerlessWebsocketsPlugin);

    this.serverless = serverless;
    this.options = options;
    this.provider = this.serverless.getProvider('aws');
    this.functions = []; // to be filled later...

    this.hooks = {
      'after:deploy:deploy': this.deployWebsockets.bind(this),
      // todo change
      'after:remove:remove': this.removeWebsockets.bind(this),
      'after:info:info': this.displayWebsockets.bind(this)
    };
  }

  _createClass(ServerlessWebsocketsPlugin, [{
    key: "getWebsocketApiName",
    value: function getWebsocketApiName() {
      if (this.serverless.service.provider.websocketApiName && is(String, this.serverless.service.provider.websocketApiName)) {
        return `${this.serverless.service.provider.websocketApiName}`;
      }

      return `${this.serverless.service.service}-${this.provider.getStage()}-websockets-api`;
    }
  }, {
    key: "getWebsocketApiRouteSelectionExpression",
    value: function getWebsocketApiRouteSelectionExpression() {
      if (this.serverless.service.provider.websocketApiRouteSelectionExpression && is(String, this.serverless.service.provider.websocketApiRouteSelectionExpression)) {
        return `${this.serverless.service.provider.websocketApiRouteSelectionExpression}`;
      }

      return `$request.body.action`;
    }
  }, {
    key: "getWebsocketUrl",
    value: function getWebsocketUrl() {
      return `wss://${this.apiId}.execute-api.${this.region}.amazonaws.com/${this.stage}/`;
    }
  }, {
    key: "init",
    value: function init() {
      this.apiName = this.getWebsocketApiName();
      this.routeSelectionExpression = this.getWebsocketApiRouteSelectionExpression();
      this.stage = this.provider.getStage();
      this.region = this.provider.getRegion();
    }
  }, {
    key: "deployWebsockets",
    value: function () {
      var _deployWebsockets = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee() {
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                this.init();
                _context.next = 3;
                return this.prepareFunctions();

              case 3:
                if (!(!is(Object, this.serverless.service.functions) || keys(this.serverless.service.functions).length === 0 || isEmpty(this.functions))) {
                  _context.next = 5;
                  break;
                }

                return _context.abrupt("return");

              case 5:
                this.serverless.cli.log(`Deploying Websockets API named "${this.apiName}"...`);
                _context.next = 8;
                return this.createApi();

              case 8:
                _context.next = 10;
                return this.createRoutes();

              case 10:
                _context.next = 12;
                return this.createDeployment();

              case 12:
                this.serverless.cli.log(`Websockets API named "${this.apiName}" with ID "${this.apiId}" has been deployed.`);
                this.serverless.cli.log(`  Websocket URL: ${this.getWebsocketUrl()}`);

              case 14:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function deployWebsockets() {
        return _deployWebsockets.apply(this, arguments);
      }

      return deployWebsockets;
    }()
  }, {
    key: "prepareFunctions",
    value: function () {
      var _prepareFunctions = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee2() {
        var _this = this;

        var res, outputs;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.provider.request('CloudFormation', 'describeStacks', {
                  StackName: this.provider.naming.getStackName()
                });

              case 2:
                res = _context2.sent;
                outputs = res.Stacks[0].Outputs;
                keys(this.serverless.service.functions || {}).map(function (name) {
                  var func = _this.serverless.service.functions[name];

                  if (func.events && func.events.find(function (event) {
                    return event.websocket;
                  })) {
                    // find the arn of this function in the list of outputs...
                    var outputKey = _this.provider.naming.getLambdaVersionOutputLogicalId(name);

                    var arn = outputs.find(function (output) {
                      return output.OutputKey === outputKey;
                    }).OutputValue; // get list of route keys configured for this function

                    var routes = map(function (e) {
                      return e.websocket.routeKey;
                    }, filter(function (e) {
                      return e.websocket;
                    }, func.events));
                    var fn = {
                      arn: arn,
                      routes
                    };

                    _this.functions.push(fn);
                  }
                });

              case 5:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function prepareFunctions() {
        return _prepareFunctions.apply(this, arguments);
      }

      return prepareFunctions;
    }()
  }, {
    key: "getApi",
    value: function () {
      var _getApi = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee3() {
        var _this2 = this;

        var apis, websocketApi;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.provider.request('ApiGatewayV2', 'getApis', {});

              case 2:
                apis = _context3.sent;
                // todo what if existing api is not valid websocket api? or non existent?
                websocketApi = apis.Items.find(function (api) {
                  return api.Name === _this2.apiName;
                });
                this.apiId = websocketApi ? websocketApi.ApiId : null;
                return _context3.abrupt("return", this.apiId);

              case 6:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getApi() {
        return _getApi.apply(this, arguments);
      }

      return getApi;
    }()
  }, {
    key: "createApi",
    value: function () {
      var _createApi = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee4() {
        var params, res;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.getApi();

              case 2:
                if (this.apiId) {
                  _context4.next = 8;
                  break;
                }

                params = {
                  Name: this.apiName,
                  ProtocolType: 'WEBSOCKET',
                  RouteSelectionExpression: this.routeSelectionExpression
                };
                _context4.next = 6;
                return this.provider.request('ApiGatewayV2', 'createApi', params);

              case 6:
                res = _context4.sent;
                this.apiId = res.ApiId;

              case 8:
                return _context4.abrupt("return", this.apiId);

              case 9:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function createApi() {
        return _createApi.apply(this, arguments);
      }

      return createApi;
    }()
  }, {
    key: "createIntegration",
    value: function () {
      var _createIntegration = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee5(arn) {
        var params, res;
        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                params = {
                  ApiId: this.apiId,
                  IntegrationMethod: 'POST',
                  IntegrationType: 'AWS_PROXY',
                  IntegrationUri: `arn:aws:apigateway:${this.region}:lambda:path/2015-03-31/functions/${arn}/invocations` // integration creation overwrites existing identical integration
                  // so we don't need to check for existance

                };
                _context5.next = 3;
                return this.provider.request('ApiGatewayV2', 'createIntegration', params);

              case 3:
                res = _context5.sent;
                return _context5.abrupt("return", res.IntegrationId);

              case 5:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function createIntegration(_x) {
        return _createIntegration.apply(this, arguments);
      }

      return createIntegration;
    }()
  }, {
    key: "addPermission",
    value: function () {
      var _addPermission = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee6(arn) {
        var functionName, accountId, region, params;
        return regeneratorRuntime.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                functionName = arn.split(':')[6];
                accountId = arn.split(':')[4];
                region = arn.split(':')[3];
                params = {
                  Action: 'lambda:InvokeFunction',
                  FunctionName: arn,
                  Principal: 'apigateway.amazonaws.com',
                  SourceArn: `arn:aws:execute-api:${region}:${accountId}:${this.apiId}/*/*`,
                  StatementId: `${functionName}-websocket`
                };
                return _context6.abrupt("return", this.provider.request('Lambda', 'addPermission', params).catch(function (e) {
                  if (e.providerError.code !== 'ResourceConflictException') {
                    throw e;
                  }
                }));

              case 5:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function addPermission(_x2) {
        return _addPermission.apply(this, arguments);
      }

      return addPermission;
    }()
  }, {
    key: "createRoute",
    value: function () {
      var _createRoute = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee7(integrationId, route) {
        var params;
        return regeneratorRuntime.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                params = {
                  ApiId: this.apiId,
                  RouteKey: route,
                  Target: `integrations/${integrationId}`
                };
                return _context7.abrupt("return", this.provider.request('ApiGatewayV2', 'createRoute', params).catch(function (e) {
                  if (e.providerError.code !== 'ConflictException') {
                    throw e;
                  }
                }));

              case 2:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function createRoute(_x3, _x4) {
        return _createRoute.apply(this, arguments);
      }

      return createRoute;
    }()
  }, {
    key: "clearRoutes",
    value: function () {
      var _clearRoutes = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee8() {
        var _this3 = this;

        var res;
        return regeneratorRuntime.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this.provider.request('ApiGatewayV2', 'getRoutes', {
                  ApiId: this.apiId
                });

              case 2:
                res = _context8.sent;
                return _context8.abrupt("return", all(map(function (route) {
                  return _this3.provider.request('ApiGatewayV2', 'deleteRoute', {
                    ApiId: _this3.apiId,
                    RouteId: route.RouteId
                  });
                }, res.Items)));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function clearRoutes() {
        return _clearRoutes.apply(this, arguments);
      }

      return clearRoutes;
    }()
  }, {
    key: "createRoutes",
    value: function () {
      var _createRoutes = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee10() {
        var _this4 = this;

        var integrationsPromises;
        return regeneratorRuntime.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this.clearRoutes();

              case 2:
                integrationsPromises = map(
                /*#__PURE__*/
                function () {
                  var _ref = _asyncToGenerator(
                  /*#__PURE__*/
                  regeneratorRuntime.mark(function _callee9(fn) {
                    var integrationId, routesPromises;
                    return regeneratorRuntime.wrap(function _callee9$(_context9) {
                      while (1) {
                        switch (_context9.prev = _context9.next) {
                          case 0:
                            _context9.next = 2;
                            return _this4.createIntegration(fn.arn);

                          case 2:
                            integrationId = _context9.sent;
                            _context9.next = 5;
                            return _this4.addPermission(fn.arn);

                          case 5:
                            routesPromises = map(function (route) {
                              return _this4.createRoute(integrationId, route);
                            }, fn.routes);
                            return _context9.abrupt("return", all(routesPromises));

                          case 7:
                          case "end":
                            return _context9.stop();
                        }
                      }
                    }, _callee9, this);
                  }));

                  return function (_x5) {
                    return _ref.apply(this, arguments);
                  };
                }(), this.functions);
                return _context10.abrupt("return", all(integrationsPromises));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function createRoutes() {
        return _createRoutes.apply(this, arguments);
      }

      return createRoutes;
    }()
  }, {
    key: "createDeployment",
    value: function () {
      var _createDeployment = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee11() {
        var _this5 = this;

        var _ref2, DeploymentId, params;

        return regeneratorRuntime.wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this.provider.request('ApiGatewayV2', 'createDeployment', {
                  ApiId: this.apiId
                });

              case 2:
                _ref2 = _context11.sent;
                DeploymentId = _ref2.DeploymentId;
                params = {
                  ApiId: this.apiId,
                  StageName: this.stage,
                  DeploymentId
                };
                return _context11.abrupt("return", this.provider.request('ApiGatewayV2', 'updateStage', params).catch(function (e) {
                  if (e.providerError.code === 'NotFoundException') {
                    return _this5.provider.request('ApiGatewayV2', 'createStage', params);
                  }
                }));

              case 6:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function createDeployment() {
        return _createDeployment.apply(this, arguments);
      }

      return createDeployment;
    }()
  }, {
    key: "removeWebsockets",
    value: function () {
      var _removeWebsockets = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee12() {
        return regeneratorRuntime.wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                this.init();
                _context12.next = 3;
                return this.getApi();

              case 3:
                if (this.apiId) {
                  _context12.next = 5;
                  break;
                }

                return _context12.abrupt("return");

              case 5:
                this.serverless.cli.log(`Removing Websockets API named "${this.apiName}" with ID "${this.apiId}"`);
                return _context12.abrupt("return", this.provider.request('ApiGatewayV2', 'deleteApi', {
                  ApiId: this.apiId
                }));

              case 7:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function removeWebsockets() {
        return _removeWebsockets.apply(this, arguments);
      }

      return removeWebsockets;
    }()
  }, {
    key: "displayWebsockets",
    value: function () {
      var _displayWebsockets = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee13() {
        var _this6 = this;

        var baseUrl, routes;
        return regeneratorRuntime.wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                this.init();
                _context13.next = 3;
                return this.prepareFunctions();

              case 3:
                if (!isEmpty(this.functions)) {
                  _context13.next = 5;
                  break;
                }

                return _context13.abrupt("return");

              case 5:
                _context13.next = 7;
                return this.getApi();

              case 7:
                baseUrl = this.getWebsocketUrl();
                routes = flatten(map(function (fn) {
                  return fn.routes;
                }, this.functions));
                this.serverless.cli.consoleLog(chalk.yellow('WebSockets:'));
                this.serverless.cli.consoleLog(`  ${chalk.yellow('Base URL:')} ${baseUrl}`);
                this.serverless.cli.consoleLog(chalk.yellow('  Routes:'));
                map(function (route) {
                  return _this6.serverless.cli.consoleLog(`    - ${baseUrl}${route}`);
                }, routes);

              case 13:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function displayWebsockets() {
        return _displayWebsockets.apply(this, arguments);
      }

      return displayWebsockets;
    }()
  }]);

  return ServerlessWebsocketsPlugin;
}();

module.exports = ServerlessWebsocketsPlugin;
